pkg/cgroups/cgroupcontrol.go

```diff
@@ -0,0 +1,221 @@
+func (c Controller) Path() string {
+	return path.Join(mountDir, c.String())
+}
+
+// RelPath returns the relative path of the given controller.
+func (c Controller) RelPath() string {
+	return c.String()
+}
+
+// Group returns the given group for the controller.
+func (c Controller) Group(group string) Group {
+	return Group(path.Join(mountDir, c.String(), group))
+}
+
+// AsGroup returns the group for the given absolute directory path.
+func AsGroup(absDir string) Group {
+	return Group(absDir)
+}
+
+// Controller returns the controller for the group.
+func (g Group) Controller() Controller {
+	relPath := strings.TrimPrefix(string(g), mountDir+"/")
+	split := strings.SplitN(relPath, "/", 2)
+	if len(split) > 0 {
+		return controllerDirs[split[0]]
+	}
+	return UnknownController
+}
+
+// GetTasks reads the pids of threads currently assigned to the group.
+func (g Group) GetTasks() ([]string, error) {
+	return g.readPids(Tasks)
+}
+
+// GetProcesses reads the pids of processes currently assigned to the group.
+func (g Group) GetProcesses() ([]string, error) {
+	return g.readPids(Procs)
+}
+
+// AddTasks writes the given thread pids to the group.
+func (g Group) AddTasks(pids ...string) error {
+	return g.writePids(Tasks, pids...)
+}
+
+// AddProcesses writes the given process pids to the group.
+func (g Group) AddProcesses(pids ...string) error {
+	return g.writePids(Procs, pids...)
+}
+
+// Write writes the formatted data to the groups entry.
+func (g Group) Write(entry, format string, args ...interface{}) error {
+	entryPath := path.Join(string(g), entry)
+	f, err := os.OpenFile(entryPath, os.O_WRONLY, 0644)
+	if err != nil {
+		return g.errorf("%q: failed to open: %v", entry, err)
+	}
+	defer f.Close()
+
+	data := fmt.Sprintf(format, args...)
+	if _, err := f.Write([]byte(data)); err != nil {
+		return g.errorf("%q: failed to write %q: %v", entry, data, err)
+	}
+
+	return nil
+}
+
+// readPids reads pids from a cgroup's tasks or procs entry.
+func (g Group) readPids(entry string) ([]string, error) {
+	var pids []string
+
+	pidFile := path.Join(string(g), entry)
+
+	f, err := os.OpenFile(pidFile, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, g.errorf("failed to open %q: %v", entry, err)
+	}
+	defer f.Close()
+
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		pids = append(pids, s.Text())
+	}
+	if s.Err() != nil {
+		return nil, g.errorf("failed to read %q: %v", entry, err)
+	}
+
+	return pids, nil
+}
+
+// writePids writes pids to a cgroup's tasks or procs entry.
+func (g Group) writePids(entry string, pids ...string) error {
+	pidFile := path.Join(string(g), entry)
+
+	f, err := os.OpenFile(pidFile, os.O_WRONLY, 0644)
+	if err != nil {
+		return g.errorf("failed to write pids to %q: %v", pidFile, err)
+	}
+	defer f.Close()
+
+	for _, pid := range pids {
+		if _, err := f.Write([]byte(pid)); err != nil {
+			if !errors.Is(err, syscall.ESRCH) {
+				return g.errorf("failed to write pid %s to %q: %v",
+					pidFile, pid, err)
```

Comment:
Found in unit testing: the order of error string format parameters should be `pid, pidFile, err`

Reply:
Thanks, updated.